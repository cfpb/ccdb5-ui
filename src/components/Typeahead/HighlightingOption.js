import PropTypes from 'prop-types';
import React from 'react';

export const HighlightingOption = ({ label, position, value }) => {
  if (position < 0) {
    return <span>{label}</span>;
  }

  const start = label.substring(0, position);
  const match = label.substr(position, value.length);
  const end = label.substring(position + value.length);

  return (
    <span>
      {start}
      <b>{match}</b>
      {end}
    </span>
  );
};

HighlightingOption.propTypes = {
  label: PropTypes.string.isRequired,
  position: PropTypes.number.isRequired,
  value: PropTypes.string.isRequired,
};

export default HighlightingOption;
