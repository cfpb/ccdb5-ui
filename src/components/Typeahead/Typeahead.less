.typeahead {
  input {
    box-sizing: border-box;
    width: 100%;
    &:focus {
      border: solid 2px var(--pacific);
    }
  }
  .m-btn-inside-input {
    .a-btn {
      position: absolute;
      right: 1em;
      top: ~'calc(50% - 8px)';
    }
    .cf-icon-svg {
      padding-right: @gutter-minimum * 0.8;
    }
  }
  &-selector {
    position: relative;

    // z-index: 99;

    ul {
      padding-left: 0;
      border: solid 2px var(--pacific);
      border-top: 0;

      position: absolute;
      top: 0;
      right: 0;
      left: 0;

      li.footer {
        padding: @gutter-normal;
        border-top: solid 1px var(--gray-40);
        margin: 0;
        background-color: var(--pacific-20);
        cursor: default;
        font-size: @size-vi;
        font-style: italic;
        list-style-type: none;
        text-align: center;
      }
    }

    input {
      &:extend(.a-text-input);
      &:focus {
        &:extend(.a-text-input:focus);
      }
      padding-left: 1.78571429em !important;
      padding-bottom: 0.28em !important;
    }
  }

  .a-btn.a-btn__link {
    color: var(--gray-80);
  }

  .rbt-menu.dropdown-menu.show {
    z-index: 100;
    overflow: visible !important;
  }

  .dropdown-item.disabled {
    color: black;
  }

  .rbt-input-hint {
    color: transparent !important;
  }

  .dropdown-item.rbt-menu-pagination-option {
    display: block;
    position: relative;
    padding: @gutter-normal;
    border-top: solid 1px var(--gray-40);
    margin: 0;
    background-color: var(--pacific-20);
    cursor: default;
    font-size: @size-vi;
    font-style: italic;
    list-style-type: none;
    text-align: center;
    width: auto;
  }

  li.typeahead-option {
    padding: @gutter-normal;
    border-top: solid 1px var(--gray-40);
    margin: 0;
    background-color: var(--white);
    cursor: default;
    list-style-type: none;
    &:hover,
    &.selected {
      border-top: solid 2px var(--pacific);
      border-bottom: solid 2px var(--pacific);
      background-color: var(--gray-10);
    }
  }
}

.input-contains-label_before {
  position: absolute;
  left: 0.85714286em;
  top: 0.5em;
  cursor: pointer;
  font-size: 0.875em;
  ~ input[type='text'] {
    padding-left: 1.78571429em;
    padding-bottom: 0.28em;
  }
  &__search .cf-icon-svg {
    position: relative;
    top: 2px;
  }
  .cf-icon-svg {
    z-index: 99;
  }
}
