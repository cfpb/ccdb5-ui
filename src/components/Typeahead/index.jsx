// Adapted from https://github.com/fmoo/react-typeahead

import './Typeahead.less'
import * as keys from '../../constants'
import { bindAll, debounce } from '../../utils'
import iconMap from '../iconMap'
import PropTypes from 'prop-types'
import React from 'react'
import Selector from './Selector'

// ----------------------------------------------------------------------------
// Usage Mode

// The user can enter any text or choose one of the drop down options
export const MODE_OPEN = 'OPEN'

// The user is only allowed to choose one of the drop down options
export const MODE_CLOSED = 'CLOSED'

// ----------------------------------------------------------------------------
// State Machine

const ERROR = 'ERROR'
const EMPTY = 'EMPTY'
const ACCUM = 'ACCUM'
const WAITING = 'WAITING'
const NO_RESULTS = 'NO_RESULTS'
const RESULTS = 'RESULTS'
const TOO_MANY = 'TOO_MANY'
const CHOSEN = 'CHOSEN'

export const nextStateFromValue = ( value, props ) => {
  let phase = value ? WAITING : EMPTY
  if ( value && value.length < props.minLength ) {
    phase = ACCUM
  }

  return {
    inputValue: value,
    phase,
    searchResults: [],
    selectedIndex: -1
  }
}

export const nextStateFromOptions = ( options, props ) => {
  let phase = RESULTS
  if ( !options || options.length === 0 ) {
    phase = NO_RESULTS
  } else if ( options.length > props.maxVisible ) {
    phase = TOO_MANY
  }

  return {
    phase,
    searchResults: options,
    selectedIndex: -1
  }
}

// ----------------------------------------------------------------------------
// Class

export default class Typeahead extends React.Component {
  constructor( props ) {
    super( props )
    this.state = {
      ...nextStateFromValue( props.value, props ),
      focused: false
    }

    // Bindings
    bindAll( this, [
      '_callForOptions',
      '_closedChooseIndex', '_closedKeyCancel',
      '_renderError', '_renderEmpty', '_renderNoResults', '_renderResults',
      '_renderTooManyResults', '_renderWaiting',
      '_onBlur', '_onFocus', '_onKeyDown', '_onOptionsError',
      '_openClear', '_openKeyCancel', '_openKeyEnter',
      '_selectOption', '_setOptions',
      '_valueUpdated'
    ] )

    // Render/Phase Map
    this.renderMap = {
      ERROR: this._renderError,
      EMPTY: this._renderEmpty,
      ACCUM: this._renderEmpty,
      WAITING: this._renderWaiting,
      NO_RESULTS: this._renderNoResults,
      RESULTS: this._renderResults,
      TOO_MANY: this._renderTooManyResults,
      CHOSEN: this._renderEmpty
    }

    // Key/function map
    this.keyMap = {}

    if ( props.mode === MODE_OPEN ) {
      this.keyMap[keys.VK_ESCAPE] = this._openKeyCancel
      this.keyMap[keys.VK_UP] = this._openNav.bind( this, -1 )
      this.keyMap[keys.VK_DOWN] = this._openNav.bind( this, 1 )
      this.keyMap[keys.VK_ENTER] = this._openKeyEnter
      this.keyMap[keys.VK_RETURN] = this._openKeyEnter
      this.keyMap[keys.VK_TAB] = this._closedChooseIndex

      // In open mode, just hide the fact that no typeahead results match
      this.renderMap[NO_RESULTS] = this._renderEmpty
    } else {
      this.keyMap[keys.VK_ESCAPE] = this._closedKeyCancel
      this.keyMap[keys.VK_UP] = this._closedNav.bind( this, -1 )
      this.keyMap[keys.VK_DOWN] = this._closedNav.bind( this, 1 )
      this.keyMap[keys.VK_ENTER] = this._closedChooseIndex
      this.keyMap[keys.VK_RETURN] = this._closedChooseIndex
      this.keyMap[keys.VK_TAB] = this._closedChooseIndex
    }

    this.search = props.debounceWait ?
      debounce( this._callForOptions, props.debounceWait ) :
      this._callForOptions
  }

  // eslint-disable-next-line camelcase
  UNSAFE_componentWillReceiveProps( nextProps ) {
    this.setState( nextStateFromValue( nextProps.value, nextProps ) )
  }

  componentDidUpdate() {
    this.search()
  }

  render() {
    const clearAction = this.props.mode === MODE_OPEN ?
      this._openClear :
      this._closedKeyCancel

    const clear = <button className="a-btn a-btn__link"
                          onClick={ clearAction }>
                      { iconMap.getIcon( 'delete' ) }
                      Clear
                  </button>

    return (
      <section className={`typeahead ${ this.props.className }`}
               onBlur={this._onBlur}
               onFocus={this._onFocus}>
        <div className="m-btn-inside-input input-contains-label">
          <div className="input-contains-label_before
                          input-contains-label_before__search">
            { iconMap.getIcon( 'search' ) }
          </div>

          <label className="u-visually-hidden" htmlFor={ this.props.htmlId }>
            { this.props.ariaLabel }
          </label>
          <input type="text"
                 autoComplete="off"
                 className="a-text-input"
                 disabled={this.props.disabled}
                 id={ this.props.htmlId }
                 onChange={this._valueUpdated}
                 onKeyDown={this._onKeyDown}
                 placeholder={this.props.placeholder}
                 value={this.state.inputValue}
                 {...this.props.textBoxProps}
          />
          { this.state.inputValue ? clear : null }
         </div>
        { this.state.focused ? this.renderMap[this.state.phase]() : null }
      </section>
    )
  }

  // --------------------------------------------------------------------------
  // Event Methods

  _onBlur() {
    this.setState( { focused: false } )
  }

  _onFocus() {
    this.setState( { focused: true } )
  }

  _onKeyDown( event ) {
    const handler = this.keyMap[event.which]
    if ( handler ) {
      handler( event )
    }
  }

  _valueUpdated( event ) {
    const nextState = nextStateFromValue( event.target.value, this.props )
    this.setState( nextState )
  }

  // --------------------------------------------------------------------------
  // Search Methods

  _calculateNewIndex( delta ) {
    const max = Math.min(
      this.props.maxVisible, this.state.searchResults.length
    )
    if ( max === 0 ) {
      return -1
    }

    // Clamp the new index
    let newIndex = this.state.selectedIndex + delta
    if ( newIndex < 0 ) {
      newIndex = 0
    }
    if ( newIndex >= max ) {
      newIndex = max - 1
    }

    return newIndex
  }

  _callForOptions() {
    if ( this.state.phase !== WAITING ) {
      return
    }

    const value = this.state.inputValue
    const returned = this.props.onInputChange( value )

    // Did the callback produce a promise?
    if ( typeof returned.then === 'function' ) {
      returned.then(
        options => this._setOptions( options ),
        error => this._onOptionsError( error )
      )
    } else {
      this._setOptions( returned )
    }
  }

  _onOptionsError() {
    this.setState( { phase: ERROR } )
  }

  _setOptions( options ) {
    const nextState = nextStateFromOptions( options, this.props )
    this.setState( nextState )
  }

  _selectOption( index ) {
    this.props.onOptionSelected( this.state.searchResults[index] )
    const nextState = {
      phase: CHOSEN,
      searchResults: [],
      selectedIndex: -1
    }

    if ( this.props.mode === MODE_CLOSED ) {
      nextState.inputValue = ''
    }

    this.setState( nextState )
  }

  // --------------------------------------------------------------------------
  // Key Helpers

  _closedChooseIndex( event ) {
    if ( this.state.searchResults.length === 0 ) {
      return;
    }

    let idx = this.state.selectedIndex
    if ( idx === -1 ) {
      idx = 0
    }

    this._selectOption( idx )
    event.preventDefault()
  }

  _closedKeyCancel( event ) {
    event.preventDefault()
    this.setState( nextStateFromValue( '', this.props ) )
  }

  _closedNav( delta, event ) {
    event.preventDefault()
    const newIndex = this._calculateNewIndex( delta )
    if ( newIndex >= 0 ) {
      this.setState( { selectedIndex: newIndex } )
    }
  }

  _openClear() {
    this.setState( nextStateFromValue( '', this.props ) )
    this.props.onOptionSelected( '' )
  }

  _openKeyEnter( event ) {
    event.preventDefault()
    this.props.onOptionSelected( this.state.inputValue )
    this.setState( { phase: CHOSEN } )
  }

  _openKeyCancel( event ) {
    event.preventDefault()
    this.setState( { phase: CHOSEN } )
  }

  _openNav( delta, event ) {
    event.preventDefault()
    const newIndex = this._calculateNewIndex( delta )
    if ( newIndex >= 0 ) {
      // May need an extractor callback eventually
      // For now, assume the shape of the options
      const inputValue = this.state.searchResults[newIndex].key

      this.setState( {
        selectedIndex: newIndex,
        inputValue
      } )
    }
  }

  // --------------------------------------------------------------------------
  // Render Helpers

  _renderError() {
    return <span className="error">
      There was a problem retrieving the options
    </span>
  }

  _renderEmpty() {
    return null
  }

  _renderWaiting() {
    return this.props.disableTypeahead ?
        null : <span className="waiting">waiting...</span>
  }

  _renderNoResults() {
    return <span className="no-results">No results found</span>
  }

  _renderResults() {
    return (
        <Selector options={this.state.searchResults}
                  onOptionSelected={this._selectOption}
                  renderOption={this.props.renderOption}
                  selectedIndex={this.state.selectedIndex}
        />
    )
  }

  _renderTooManyResults() {
    const subset = this.state.searchResults.slice( 0, this.props.maxVisible )
    return (
        <Selector options={subset}
                  onOptionSelected={this._selectOption}
                  renderOption={this.props.renderOption}
                  selectedIndex={this.state.selectedIndex}
                  footer="Continue typing for more results"
        />
    )
  }
}

Typeahead.propTypes = {
  ariaLabel: PropTypes.string.isRequired,
  className: PropTypes.string,
  debounceWait: PropTypes.number,
  disableTypeahead: PropTypes.bool,
  htmlId: PropTypes.string.isRequired,
  maxVisible: PropTypes.number,
  minLength: PropTypes.number,
  mode: PropTypes.oneOf( [ MODE_OPEN, MODE_CLOSED ] ).isRequired,
  onInputChange: PropTypes.func.isRequired,
  onOptionSelected: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  renderOption: PropTypes.func,
  textBoxProps: PropTypes.object,
  value: PropTypes.string
}

Typeahead.defaultProps = {
  className: '',
  debounceWait: 0,
  maxVisible: 5,
  minLength: 2,
  mode: MODE_CLOSED,
  placeholder: 'Enter your search text',
  renderOption: x => ( {
    value: x,
    component: x
  } ),
  textBoxProps: {},
  value: ''
}
