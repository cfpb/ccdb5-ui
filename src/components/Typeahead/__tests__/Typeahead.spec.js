import { mount } from 'enzyme';
import React from 'react';
import Typeahead, { MODE_OPEN } from '../index';
import * as keys from '../../../constants';

/**
 *
 * @param initalProps
 * @param removeDebounce
 */
function setupEnzyme(initalProps = {}, removeDebounce = true) {
  const props = Object.assign(
    {
      ariaLabel: 'Start typing to...',
      htmlId: 'typeahead-foo',
      onInputChange: jest.fn(() => ['alpha', 'beta', 'gamma']),
      onOptionSelected: jest.fn(),
      renderOption: jest.fn(),
    },
    initalProps
  );

  const target = mount(<Typeahead {...props} />);

  // Remove debounce
  if (removeDebounce) {
    const instance = target.instance();
    instance.search = instance._callForOptions;
  }

  return {
    props,
    target,
  };
}

describe('component::Typeahead', () => {
  describe('focus/blur', () => {
    const renderOption = (x) => ({
      value: x,
      component: x,
    });

    it('sets the current state on blur', () => {
      const { target } = setupEnzyme({ renderOption, value: 'foo' });
      target.setState({ focused: true });
      target.simulate('blur');
      expect(target.state('focused')).toEqual(false);
    });

    it('sets the state on focus', () => {
      const { target } = setupEnzyme({ renderOption, value: 'foo' });
      expect(target.state('phase')).toEqual('WAITING');
      expect(target.state('focused')).toEqual(false);
      target.simulate('focus');
      expect(target.state('focused')).toEqual(true);
    });
  });

  describe('text control', () => {
    let target, props, input;
    beforeEach(() => {
      ({ target, props } = setupEnzyme());
      input = target.find('input');
    });

    it('requires a minimum number of characters before calling for options', () => {
      input.simulate('change', { target: { value: 'b' } });
      expect(target.state('inputValue')).toEqual('b');
      expect(props.onInputChange).not.toHaveBeenCalled();
    });

    it('calls for options when the threshhold has been reached', () => {
      input.simulate('change', { target: { value: 'bar' } });
      expect(target.state('inputValue')).toEqual('bar');
      expect(props.onInputChange).toHaveBeenCalledWith('bar');
    });
  });

  describe('asynchronous options', () => {
    let target, props, input;
    let fakePromise, onSuccess, onFail;
    beforeEach(() => {
      ({ target, props } = setupEnzyme());
      input = target.find('input');

      fakePromise = {
        then: (x, y) => {
          onSuccess = x;
          onFail = y;
        },
      };
      props.onInputChange.mockImplementation(() => fakePromise);
    });

    it('detects a promise', () => {
      input.simulate('change', { target: { value: 'bar' } });
      expect(props.onInputChange).toHaveBeenCalledWith('bar');
      expect(target.state('inputValue')).toEqual('bar');
      expect(target.state('phase')).toEqual('WAITING');
    });

    it('sets the options when the promise is resolved', () => {
      input.simulate('change', { target: { value: 'bar' } });
      expect(props.onInputChange).toHaveBeenCalledWith('bar');
      expect(target.state('phase')).toEqual('WAITING');
      onSuccess(['a', 'b']);
      expect(target.state('phase')).toEqual('RESULTS');
    });

    it('enters the error state when the promise is rejected', () => {
      input.simulate('change', { target: { value: 'bar' } });
      expect(props.onInputChange).toHaveBeenCalledWith('bar');
      expect(target.state('phase')).toEqual('WAITING');
      onFail('oops');
      expect(target.state('phase')).toEqual('ERROR');
    });
  });

  describe('keyboard events in CLOSED mode', () => {
    let fixture, target, props, input;
    beforeEach(() => {
      ({ target, props } = setupEnzyme());
      input = target.find('input');
      fixture = {
        preventDefault: jest.fn(),
      };
    });

    it('ignores unknown keys', () => {
      fixture.which = 999;
      input.simulate('keydown', fixture);
      expect(fixture.preventDefault).not.toHaveBeenCalled();
    });

    it('resets the control when "ESC" is pressed', () => {
      input.simulate('change', { target: { value: 'bar' } });
      expect(target.state('phase')).toEqual('RESULTS');

      fixture.which = keys.VK_ESCAPE;
      input.simulate('keydown', fixture);

      expect(target.state('inputValue')).toEqual('');
      expect(target.state('phase')).toEqual('EMPTY');
      expect(fixture.preventDefault).toHaveBeenCalled();
    });

    describe('"DOWN" key', () => {
      it('has no effect when there are no options', () => {
        fixture.which = keys.VK_DOWN;
        input.simulate('keydown', fixture);

        expect(target.state('selectedIndex')).toEqual(-1);
        expect(fixture.preventDefault).toHaveBeenCalled();
      });

      it('selects the first option when there is no selection', () => {
        input.simulate('change', { target: { value: 'bar' } });
        fixture.which = keys.VK_DOWN;
        input.simulate('keydown', fixture);

        expect(target.state('selectedIndex')).toEqual(0);
      });

      it('has no effect when the last item is selected', () => {
        target.instance().setState({
          selectedIndex: 2,
          searchResults: ['foo', 'bar', 'baz'],
        });

        fixture.which = keys.VK_DOWN;
        input.simulate('keydown', fixture);

        expect(target.state('selectedIndex')).toEqual(2);
      });
    });

    describe('"UP" key', () => {
      it('has no effect when the first item is selected', () => {
        target.instance().setState({
          selectedIndex: 0,
          searchResults: ['foo', 'bar', 'baz'],
        });

        fixture.which = keys.VK_UP;
        input.simulate('keydown', fixture);

        expect(target.state('selectedIndex')).toEqual(0);
      });
    });

    describe('ENTER/TAB', () => {
      it('has no effect when there are no options', () => {
        fixture.which = keys.VK_ENTER;
        input.simulate('keydown', fixture);
        expect(props.onOptionSelected).not.toHaveBeenCalled();
      });

      it('chooses the first option when there is no selection', () => {
        input.simulate('change', { target: { value: 'bar' } });
        fixture.which = keys.VK_TAB;
        input.simulate('keydown', fixture);

        expect(props.onOptionSelected).toHaveBeenCalledWith('alpha');
      });

      it('selects the highlighted option when "TAB" is pressed', () => {
        input.simulate('change', { target: { value: 'bar' } });
        target.instance().setState({
          selectedIndex: 1,
        });
        fixture.which = keys.VK_TAB;
        input.simulate('keydown', fixture);

        expect(props.onOptionSelected).toHaveBeenCalledWith('beta');
      });
    });
  });

  describe('keyboard events in OPEN mode', () => {
    let fixture, target, props, input;
    beforeEach(() => {
      ({ target, props } = setupEnzyme({
        mode: MODE_OPEN,
        onInputChange: jest.fn(() => [{ key: 'alpha' }, { key: 'beta' }]),
      }));

      input = target.find('input');
      fixture = {
        preventDefault: jest.fn(),
      };

      input.simulate('change', { target: { value: 'bar' } });
      expect(target.state('phase')).toEqual('RESULTS');
    });

    afterEach(() => {
      expect(fixture.preventDefault).toHaveBeenCalled();
    });

    it('hides the drop down when "ESC" is pressed', () => {
      fixture.which = keys.VK_ESCAPE;
      input.simulate('keydown', fixture);

      expect(target.state('inputValue')).toEqual('bar');
      expect(target.state('phase')).toEqual('CHOSEN');
    });

    describe('arrow keys', () => {
      it('sets the text box value to the selected value', () => {
        fixture.which = keys.VK_DOWN;
        input.simulate('keydown', fixture);

        expect(target.state('selectedIndex')).toEqual(0);
        expect(target.state('inputValue')).toEqual('alpha');
      });

      it('has no effect when there are no results', () => {
        target.instance()._calculateNewIndex = jest.fn(() => -1);

        fixture.which = keys.VK_DOWN;
        input.simulate('keydown', fixture);

        expect(target.state('selectedIndex')).toEqual(-1);
        expect(target.state('inputValue')).toEqual('bar');
      });
    });

    describe('ENTER/TAB', () => {
      it('selects the highlighted option', () => {
        target.instance().setState({
          selectedIndex: 1,
        });
        fixture.which = keys.VK_TAB;
        input.simulate('keydown', fixture);

        expect(props.onOptionSelected).toHaveBeenCalledWith({ key: 'beta' });
        expect(target.state('inputValue')).toEqual('bar');
        expect(target.state('phase')).toEqual('CHOSEN');
        expect(target.state('searchResults')).toEqual([]);
        expect(target.state('selectedIndex')).toEqual(-1);
      });

      it('uses the text and not the highlighted option', () => {
        target.instance().setState({
          selectedIndex: 1,
          inputValue: 'be',
        });
        fixture.which = keys.VK_ENTER;
        input.simulate('keydown', fixture);

        expect(props.onOptionSelected).toHaveBeenCalledWith('be');
        expect(target.state('inputValue')).toEqual('be');
        expect(target.state('phase')).toEqual('CHOSEN');
      });
    });
  });

  describe('clear button in OPEN mode', () => {
    let target, props, button;
    beforeEach(() => {
      ({ target, props } = setupEnzyme({
        mode: MODE_OPEN,
        value: 'bar',
      }));

      button = target.find('button');
    });

    it('clears the value and triggers a search', () => {
      button.simulate('click');
      expect(props.onOptionSelected).toHaveBeenCalledWith('');
    });
  });
});
