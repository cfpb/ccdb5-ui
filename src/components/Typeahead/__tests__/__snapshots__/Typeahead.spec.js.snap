// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::Typeahead render phases renders the ACCUM phase 1`] = `
<section
  className="typeahead "
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    className="m-btn-inside-input input-contains-label"
  >
    <div
      className="input-contains-label_before input-contains-label_before__search"
    >
      <svg
        className="cf-icon-svg"
        viewBox="0 0 880.6 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M860.1 879.2L645.7 664.8c90.8-136.8 76-323-44.6-443.5-137.5-137.5-360.5-137.5-498 0s-137.5 360.5 0 498c118.5 118.4 303.9 137 443.5 44.6L761 978.3c27.3 27.3 71.7 27.3 99 0s27.4-71.8.1-99.1zm-508-116.9C191.4 762 61.3 631.5 61.6 470.8c.3-160.7 130.8-290.8 291.5-290.5s290.8 130.8 290.5 291.5c-.2 118.2-71.9 224.6-181.5 269.1-34.9 14.2-72.3 21.5-110 21.4z"
        />
      </svg>
    </div>
    <label
      className="u-visually-hidden"
      htmlFor="typeahead-foo"
    >
      Start typing to...
    </label>
    <input
      autoComplete="off"
      className="a-text-input"
      id="typeahead-foo"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter your search text"
      type="text"
      value="i"
    />
    <button
      className="a-btn a-btn__link"
      onClick={[Function]}
    >
      <svg
        className="cf-icon-svg"
        viewBox="0 0 718.9 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M451.4 613.7l248.1-248.1c25.6-25.1 26-66.3.8-91.9s-66.3-26-91.9-.8l-.8.8-248.1 248.1-248.1-248.1c-25.4-25.4-66.5-25.4-91.9 0s-25.4 66.5 0 91.9l248.1 248.1L19.5 861.8c-25.6 25.1-26 66.3-.8 91.9s66.3 26 91.9.8l.8-.8 248.1-248.1 248.1 248.1c25.4 25.4 66.5 25.4 91.9 0s25.4-66.5 0-91.9L451.4 613.7z"
        />
      </svg>
      Clear
    </button>
  </div>
</section>
`;

exports[`component::Typeahead render phases renders the EMPTY phase 1`] = `
<section
  className="typeahead "
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    className="m-btn-inside-input input-contains-label"
  >
    <div
      className="input-contains-label_before input-contains-label_before__search"
    >
      <svg
        className="cf-icon-svg"
        viewBox="0 0 880.6 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M860.1 879.2L645.7 664.8c90.8-136.8 76-323-44.6-443.5-137.5-137.5-360.5-137.5-498 0s-137.5 360.5 0 498c118.5 118.4 303.9 137 443.5 44.6L761 978.3c27.3 27.3 71.7 27.3 99 0s27.4-71.8.1-99.1zm-508-116.9C191.4 762 61.3 631.5 61.6 470.8c.3-160.7 130.8-290.8 291.5-290.5s290.8 130.8 290.5 291.5c-.2 118.2-71.9 224.6-181.5 269.1-34.9 14.2-72.3 21.5-110 21.4z"
        />
      </svg>
    </div>
    <label
      className="u-visually-hidden"
      htmlFor="typeahead-foo"
    >
      Start typing to...
    </label>
    <input
      autoComplete="off"
      className="a-text-input"
      id="typeahead-foo"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter your search text"
      type="text"
      value=""
    />
  </div>
</section>
`;

exports[`component::Typeahead render phases renders the ERROR phase 1`] = `
<section
  className="typeahead "
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    className="m-btn-inside-input input-contains-label"
  >
    <div
      className="input-contains-label_before input-contains-label_before__search"
    >
      <svg
        className="cf-icon-svg"
        viewBox="0 0 880.6 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M860.1 879.2L645.7 664.8c90.8-136.8 76-323-44.6-443.5-137.5-137.5-360.5-137.5-498 0s-137.5 360.5 0 498c118.5 118.4 303.9 137 443.5 44.6L761 978.3c27.3 27.3 71.7 27.3 99 0s27.4-71.8.1-99.1zm-508-116.9C191.4 762 61.3 631.5 61.6 470.8c.3-160.7 130.8-290.8 291.5-290.5s290.8 130.8 290.5 291.5c-.2 118.2-71.9 224.6-181.5 269.1-34.9 14.2-72.3 21.5-110 21.4z"
        />
      </svg>
    </div>
    <label
      className="u-visually-hidden"
      htmlFor="typeahead-foo"
    >
      Start typing to...
    </label>
    <input
      autoComplete="off"
      className="a-text-input"
      id="typeahead-foo"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter your search text"
      type="text"
      value=""
    />
  </div>
  <span
    className="error"
  >
    There was a problem retrieving the options
  </span>
</section>
`;

exports[`component::Typeahead render phases renders the NO_RESULTS phase 1`] = `
<section
  className="typeahead "
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    className="m-btn-inside-input input-contains-label"
  >
    <div
      className="input-contains-label_before input-contains-label_before__search"
    >
      <svg
        className="cf-icon-svg"
        viewBox="0 0 880.6 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M860.1 879.2L645.7 664.8c90.8-136.8 76-323-44.6-443.5-137.5-137.5-360.5-137.5-498 0s-137.5 360.5 0 498c118.5 118.4 303.9 137 443.5 44.6L761 978.3c27.3 27.3 71.7 27.3 99 0s27.4-71.8.1-99.1zm-508-116.9C191.4 762 61.3 631.5 61.6 470.8c.3-160.7 130.8-290.8 291.5-290.5s290.8 130.8 290.5 291.5c-.2 118.2-71.9 224.6-181.5 269.1-34.9 14.2-72.3 21.5-110 21.4z"
        />
      </svg>
    </div>
    <label
      className="u-visually-hidden"
      htmlFor="typeahead-foo"
    >
      Start typing to...
    </label>
    <input
      autoComplete="off"
      className="a-text-input"
      id="typeahead-foo"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter your search text"
      type="text"
      value=""
    />
  </div>
  <span
    className="no-results"
  >
    No results found
  </span>
</section>
`;

exports[`component::Typeahead render phases renders the RESULTS phase 1`] = `
<section
  className="typeahead "
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    className="m-btn-inside-input input-contains-label"
  >
    <div
      className="input-contains-label_before input-contains-label_before__search"
    >
      <svg
        className="cf-icon-svg"
        viewBox="0 0 880.6 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M860.1 879.2L645.7 664.8c90.8-136.8 76-323-44.6-443.5-137.5-137.5-360.5-137.5-498 0s-137.5 360.5 0 498c118.5 118.4 303.9 137 443.5 44.6L761 978.3c27.3 27.3 71.7 27.3 99 0s27.4-71.8.1-99.1zm-508-116.9C191.4 762 61.3 631.5 61.6 470.8c.3-160.7 130.8-290.8 291.5-290.5s290.8 130.8 290.5 291.5c-.2 118.2-71.9 224.6-181.5 269.1-34.9 14.2-72.3 21.5-110 21.4z"
        />
      </svg>
    </div>
    <label
      className="u-visually-hidden"
      htmlFor="typeahead-foo"
    >
      Start typing to...
    </label>
    <input
      autoComplete="off"
      className="a-text-input"
      id="typeahead-foo"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter your search text"
      type="text"
      value=""
    />
  </div>
  <div
    className="typeahead-selector"
  >
    <ul>
      <li
        className="typeahead-option body-copy"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        foo
      </li>
    </ul>
  </div>
</section>
`;

exports[`component::Typeahead render phases renders the TOO_MANY phase 1`] = `
<section
  className="typeahead "
  onBlur={[Function]}
  onFocus={[Function]}
>
  <div
    className="m-btn-inside-input input-contains-label"
  >
    <div
      className="input-contains-label_before input-contains-label_before__search"
    >
      <svg
        className="cf-icon-svg"
        viewBox="0 0 880.6 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M860.1 879.2L645.7 664.8c90.8-136.8 76-323-44.6-443.5-137.5-137.5-360.5-137.5-498 0s-137.5 360.5 0 498c118.5 118.4 303.9 137 443.5 44.6L761 978.3c27.3 27.3 71.7 27.3 99 0s27.4-71.8.1-99.1zm-508-116.9C191.4 762 61.3 631.5 61.6 470.8c.3-160.7 130.8-290.8 291.5-290.5s290.8 130.8 290.5 291.5c-.2 118.2-71.9 224.6-181.5 269.1-34.9 14.2-72.3 21.5-110 21.4z"
        />
      </svg>
    </div>
    <label
      className="u-visually-hidden"
      htmlFor="typeahead-foo"
    >
      Start typing to...
    </label>
    <input
      autoComplete="off"
      className="a-text-input"
      id="typeahead-foo"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Enter your search text"
      type="text"
      value=""
    />
  </div>
  <div
    className="typeahead-selector"
  >
    <ul>
      <li
        className="typeahead-option body-copy"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        foo
      </li>
      <li
        className="typeahead-option body-copy"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        bar
      </li>
      <li
        className="typeahead-option body-copy"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        baz
      </li>
      <li
        className="typeahead-option body-copy"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        qaz
      </li>
      <li
        className="typeahead-option body-copy"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        quux
      </li>
      <li
        className="footer"
      >
        Continue typing for more results
      </li>
    </ul>
  </div>
</section>
`;
